arangod/IResearch/GeoAnalyzer.cpp

```diff
@@ -290,93 +362,144 @@ void GeoAnalyzer::reset(std::vector<std::string>&& terms) noexcept {
   _end = _begin + _terms.size();
 }
 
-/*static*/ bool GeoJSONAnalyzer::normalize(std::string_view args,
-                                           std::string& out) {
-  return ::normalize<GeoJSONAnalyzer>(args, out);
-}
+GeoJsonAnalyzerBase::GeoJsonAnalyzerBase(
+    irs::type_info const& type, GeoJsonAnalyzerBase::OptionsBase const& options)
+    : GeoAnalyzer{type,
+                  S2Options(options.options, options.type != Type::SHAPE)},
+      _type{options.type} {}
+
+bool GeoJsonAnalyzerBase::resetImpl(std::string_view value, bool legacy,
+                                    bool legacyCovering) {
+  auto const data = slice(value);
+  if (_type != Type::POINT) {
+    if (!parseShape<Parsing::GeoJson>(data, _shape, _cache, legacy)) {
+      return false;
+    }
+  } else if (!parseShape<Parsing::OnlyPoint>(data, _shape, _cache, legacy)) {
+    return false;
+  }
 
-/*static*/ irs::analysis::analyzer::ptr GeoJSONAnalyzer::make(
-    std::string_view args) {
-  return ::make<GeoJSONAnalyzer>(args);
+  std::vector<std::string> geoTerms;
+  auto const type = _shape.type();
+  if (_type == Type::CENTROID || type == geo::ShapeContainer::Type::S2_POINT) {
+    geoTerms = _indexer.GetIndexTerms(_shape.centroid(), {});
+  } else if (legacyCovering) {
+    geoTerms = _indexer.GetIndexTerms(*_shape.region(), {});
+  } else {
+    auto cellIds = _shape.covering(_coverer);
+    // TODO(MBkkt) S2_MULTIPOINT can be handled via GetIndexTerms(point)
+    if (type == geo::ShapeContainer::Type::S2_MULTIPOINT ||
```

Comment:
Could it be here without specific type checks? Just 
`if (!_coverer.IsCanonical(cellIds)) {
  _coverer.CanonicalizeCovering(&cellIds);
}`

Reply:
IsCanonical is linear time